File: core/src/com/google/zxing/MonochromeBitmapSource.java
/*
 * Copyright 2007 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.zxing;

import com.google.zxing.common.BitArray;

/**
 * <p>Encapsulates a generic black-and-white bitmap -- a collection of pixels in two dimensions.
 * This unifies many possible representations, like AWT's <code>BufferedImage</code>.</p>
 *
 * @author srowen@google.com (Sean Owen)
 */
public interface MonochromeBitmapSource {

  /**
   * @param x horizontal offset, from left, of the pixel
   * @param y vertical offset, from top, of the pixel
   * @return true iff the pixel at (x,y) is black
   */
  boolean isBlack(int x, int y);

  /**
   * <p>Returns an entire row of black/white pixels as an array of bits, where "true" means "black".
   * This is a sort of "bulk get" operation intended to enable efficient access in
   * certain situations.</p>
   *
   * @param y vertical offset, from top, of the row of pixels
   * @param row if not null, {@link BitArray} to write pixels into. If null, a new {@link BitArray}
   * is allocated and returned.
   * @param startX horizontal offset, from left, from which to start getting pixels
   * @param getWidth number of pixels to get from the row
   * @return {@link BitArray} representing the (subset of the) row of pixels. If row parameter
   *         was not null, it is returned.
   */
  BitArray getBlackRow(int y, BitArray row, int startX, int getWidth);

  /**
   * Entirely analogous to {@link #getBlackRow(int, BitArray, int, int)} but gets a column.
   */
  BitArray getBlackColumn(int x, BitArray column, int startY, int getHeight);

  /**
   * @return height of underlying image
   */
  int getHeight();

  /**
   * @return width of underlying image
   */
  int getWidth();

  /**
   * Retrieves the luminance at the pixel x,y in the bitmap. This method is only used for estimating
   * the black point and implementing getBlackRow() - it is not meant for decoding.
   *
   * @param x The x coordinate in the image.
   * @param y The y coordinate in the image.
   * @return The luminance value between 0 and 255.
   */
  int getLuminance(int x, int y);

  /**
   * Some implementations can be much more efficient by fetching an entire row of luminance data at
   * a time. This method should be called once per row before calling getLuminance().
   *
   * @param y The row to cache.
   */
  void cacheRowForLuminance(int y);

  /**
   * Entirely analogous to {@link #cacheRowForLuminance(int)} but caches a column.
   */
  void cacheColumnForLuminance(int x);

  /**
   * <p>Estimates black point according to the given method, which is optionally parameterized by
   * a single int argument. For {@link BlackPointEstimationMethod#ROW_SAMPLING}, this
   * specifies the row to sample.</p>
   *
   * <p>The estimated value will be used in subsequent computations that rely on an estimated black
   * point.</p>
   *
   * @param method black point estimation method
   * @param argument method-specific argument
   */
  void estimateBlackPoint(BlackPointEstimationMethod method, int argument) throws ReaderException;

  /**
   * @return {@link BlackPointEstimationMethod} representing last sampling method used
   */
  BlackPointEstimationMethod getLastEstimationMethod();

  /**
   * <p>Optional operation which returns an implementation based on the same underlying
   * image, but which behaves as if the underlying image had been rotated 90 degrees
   * counterclockwise. This is useful in the context of 1D barcodes and the
   * {@link DecodeHintType#TRY_HARDER} decode hint, and is only intended to be
   * used in non-resource-constrained environments. Hence, implementations
   * of this class which are only used in resource-constrained mobile environments
   * don't have a need to implement this.</p>
   *
   * @throws IllegalArgumentException if not supported
   */
  MonochromeBitmapSource rotateCounterClockwise();

  /**
   * @return true iff rotation is supported
   * @see #rotateCounterClockwise()
   */
  boolean isRotateSupported();

}

File: cpp/core/src/MonochromeBitmapSource.cpp
/*
 *  MonochromeBitmapSource.cpp
 *  zxing
 *
 *  Created by Christian Brunschen on 12/05/2008.
 *  Copyright 2008 Google UK. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "MonochromeBitmapSource.h"
#include "BlackPointEstimator.h"
#include "ReaderException.h"
#include <limits>

#define LUMINANCE_BITS (5)
#define LUMINANCE_SHIFT (8 - LUMINANCE_BITS)
#define LUMINANCE_BUCKETS (1 << LUMINANCE_BITS)

using namespace std;

MonochromeBitmapSource::MonochromeBitmapSource() :
blackPoint_(numeric_limits<unsigned char>::max()),
lastEstimationMethod_(BlackPointEstimationMethod_None), 
lastEstimationArgument_(-1) {
}


bool MonochromeBitmapSource::isBlack(size_t x, size_t y) {
  return getPixel(x, y) < blackPoint_;
}

Ref<BitArray> MonochromeBitmapSource::getBlackRow(size_t y,
                                                  Ref<BitArray> row, 
                                                  size_t startX,
                                                  size_t getWidth) {
  Ref<BitArray> result(row ? row : new BitArray(getWidth));
  if (row) {
    result->clear();
  }
  valarray<unsigned char> pixelRow(getWidth);
  for (size_t i = 0; i < getWidth; i++) {
    pixelRow[i] = getPixel(startX + i, y);
  }
  
  // If the current decoder calculated the blackPoint based on one row, 
  // assume we're trying to decode a 1D barcode, and apply some sharpening.
  if (lastEstimationMethod_ == BlackPointEstimationMethod_RowSampling) {
    unsigned int left = pixelRow[0];
    unsigned int center = pixelRow[1];
    for (size_t i = 1; i < getWidth - 1; i++) {
      unsigned int right = pixelRow[i+1];
      // Simple -1 4 -1 box filter with a weight of 2
      unsigned int luminance = ((center << 2) - left - right) >> 1;
      if (luminance < blackPoint_) {
        result->set(i);
      }
      left = center;
      center = right;
    }
  } else {
    for (size_t i = 0; i < getWidth; i++) {
      if (pixelRow[i] < blackPoint_) {
        result->set(i);
      }
    }
  }
  
  return result;
}

void MonochromeBitmapSource::
estimateBlackPoint(BlackPointEstimationMethod method, int arg) {
  size_t width = getWidth();
  size_t height = getHeight();
  if (method != lastEstimationMethod_ || arg != lastEstimationArgument_) {
    valarray<int> histogram(LUMINANCE_BUCKETS);
    if (method == BlackPointEstimationMethod_2D) {
      size_t minDimension = width < height ? width : height;
      size_t yOffset = height == minDimension ? 0 : (height - width) >> 1;
      size_t xOffset = width == minDimension ? 0 : (width - height) >> 1;
      for (size_t n = 0; n < minDimension; n++) {
        unsigned char pixel = getPixel(xOffset + n, yOffset + n);
        histogram[pixel >> LUMINANCE_SHIFT]++;
      }
    } else if (method == BlackPointEstimationMethod_RowSampling) {
      if (arg < 0 || arg > (int)height) {
        throw new IllegalArgumentException
        ("black point estimation argument out of range");
      }
      size_t y = static_cast<size_t> (arg);
      for (size_t x = 0; x < width; x++) {
        histogram[getPixel(x, y) >> LUMINANCE_SHIFT]++;
      }
    } else {
      throw new IllegalArgumentException
      ("unknown black point estimation method");
    }
    
    blackPoint_ = BlackPointEstimator::estimate(histogram) << LUMINANCE_SHIFT;    
    lastEstimationMethod_ = method;
    lastEstimationArgument_ = arg;
  }
}

BlackPointEstimationMethod MonochromeBitmapSource::getLastEstimationMethod() {
  return BlackPointEstimationMethod_None;
}

